runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/work/refresh/ActiveWorkRefresher.java

```diff
@@ -103,5 +115,32 @@ private void invalidateStuckCommits() {
     }
   }
 
-  protected abstract void refreshActiveWork();
+  private void refreshActiveWork() {
+    Instant refreshDeadline = clock.get().minus(Duration.millis(activeWorkRefreshPeriodMillis));
+
+    Map<HeartbeatSender, Heartbeats.Builder> heartbeatsBySender = new HashMap<>();
+
+    // Aggregate the heartbeats across computations by HeartbeatSender for correct fan out.
+    for (ComputationState computationState : computations.get()) {
+      for (RefreshableWork work : computationState.getRefreshableWork(refreshDeadline)) {
+        heartbeatsBySender
+            .computeIfAbsent(work.heartbeatSender(), ignored -> Heartbeats.builder())
+            .addWork(work)
```

Comment:
the chaining confused me here since I thought the heartbeat requests were scoped to the work added

could createHeartbeatRequest be moved into HeartbeatsBuilder.addWork() and addHeartbeatRequest be removed?

Reply:
done